content/presence-occupancy/presence.textile

```diff
@@ -0,0 +1,729 @@
+
+The following is an example of a batch presence request for 2 channels:
+
+```[text]
+{
+  channel: 'channel1,channel2'
+}
+```
+
+h4(#batch-curl). Batch presence curl requests
+
+The following is an example curl request, querying the REST API directly: 
+
+```[sh]
+curl -X GET https://rest.ably.io/presence?channel=channel1,channel2 \
+    -u "{{API_KEY}}"
+```
+
+h4(#batch-responses). Batch presence responses
+
+A batch presence request can succeed, fail or partially succeed. The response returned will differ depending on the outcome. Any successful requests will return details of the clients present on each channel.
+
+The examples for each possible outcome will use the following request data:
+
+```[text]
+{
+  channel: 'channel0,channel1,channel2'
+}
+```
+
+h5(#batch-success). Batch presence success
+
+The following is an example of a successful batch presence response. The response body contains details of each client present on the channel and the @channel@ they are present on. The status code for a successful response is always @200@:
+
+```[json]
+[
+  {
+    "channel":"channel0",
+    "presence":[
+      {"clientId": "user1", "action": "1"},
+      {"clientId": "user2", "action": "1"}
+    ]
+  },
+  {
+    "channel":"channel1",
+    "presence":[]
+  },
+  {
+    "channel":"channel2",
+    "presence":[
+      {"clientId": "user2", "action": "1"},
+      {"clientId": "user3", "action": "1"}
+    ]
+  }
+]
+```
+
+h5(#batch-failure). Batch presence failure
+
+The following is an example of a batch presence failure response. The response body contains the details of the @error@, in this example that the token used for the request has expired. The status code is @401@:
+
+```[json]
+{
+  "error": {
+    "message":"Token expired",
+    "statusCode":401,
+    "code":40140
+  }
+}
+```
+
+h5(#batch-partial). Batch presence partial success
+
+The following is an example of a batch presence partial success response. The successful requests contain the details of each client present on the channel and the @channel@ they are present on. The failed request contains the @channel@ the request failed for and the details of the @error@, in this example that the credentials used didn't have the capability to query that channel. The status code for a partial success is always @400@:
+
+```[json]
+{
+  "error": {
+    "message": "Batched response includes errors",
+    "statusCode":400,
+    "code":40020
+  }
+  "batchResponse": [
+    {
+      "channel":"channel0",
+      "presence":[
+        {"clientId": "user1", "action": "1"},
+        {"clientId": "user2", "action": "1"}
+      ]
+    },
+    {
+      "channel":"channel1",
+      "presence":[]
+    },
+    {
+      "channel":"channel2",
+      "error": {
+        "message": "Given credentials do not have the required capability",
+        "statusCode": 401,
+        "code": 40160
+      }
+    }
+  ]
+}
+```
+
```

Comment:
``````suggestion
Once all requests have been completed in a batch request, a batch response is returned with three possible outcomes:

- Success := If all of the individual requests were successful then an array containing the response of each query is returned in request order.
- Failure := If the batch request itself failed before the individual requests were made, then an error response is returned with a status code and error response body. Examples of why the batch request can fail include an authorization failure or an invalid request.
- Partial success := If one or more of the individual requests failed the response body contains an error object with the error code @40020@ and a status code of @400@. The error body contains a @batchResponse@ array of each individual response in request order. The @batchResponse@ can be inspected if there is a need to know the details of each outcome. If you only need to know whether or not the batch request was completely successful then the status code is sufficient.

The examples for each possible outcome will use the following request data:

```[text]
{
  channel: 'channel0,channel1,channel2'
}
```

The following is an example of a successful batch presence response. The response body contains details of each client present on the channel and the channel they are present on. The status code for a successful response is always @200@:

```[json]
[
  {
    "channel":"channel0",
    "presence":[
      {"clientId": "user1", "action": "1"},
      {"clientId": "user2", "action": "1"}
    ]
  },
  {
    "channel":"channel1",
    "presence":[]
  },
  {
    "channel":"channel2",
    "presence":[
      {"clientId": "user2", "action": "1"},
      {"clientId": "user3", "action": "1"}
    ]
  }
]
```

The following is an example of a batch presence failure response. The response body contains the details of the @error@, in this example that the token used for the request has expired. The status code is @401@:

```[json]
{
  "error": {
    "message":"Token expired",
    "statusCode":401,
    "code":40140
  }
}
```

The following is an example of a batch presence partial success response. The successful requests contain the details of each client present on the channel, and the channel they are present on. The failed request contains the channel the request failed for and the details of the @error@, in this example that the credentials used didn't have the capability to query that channel. The status code for a partial success is always @400@:

```[json]
{
  "error": {
    "message": "Batched response includes errors",
    "statusCode":400,
    "code":40020
  }
  "batchResponse": [
    {
      "channel":"channel0",
      "presence":[
        {"clientId": "user1", "action": "1"},
        {"clientId": "user2", "action": "1"}
      ]
    },
    {
      "channel":"channel1",
      "presence":[]
    },
    {
      "channel":"channel2",
      "error": {
        "message": "Given credentials do not have the required capability",
        "statusCode": 401,
        "code": 40160
      }
    }
  ]
}
```
``````